File: colorfilter/paper/Deconvolution/shrinkv23.m
%%% Taken from the code provided by the authors of
%%% D. Krishnan, R. Fergus: "Fast Image Deconvolution using
%%% Hyper-Laplacian Priors", Proceedings of NIPS 2009. 
function w = shrinkv23(v, beta)
% solve a quartic equation
% for alpha = 2/3

epsilon = 1e-6; %% tolerance on imag part of real root  
   
k = 8/(27*beta^3);
m = ones(size(v))*k;
  
% Now use formula from
% http://en.wikipedia.org/wiki/Quartic_equation (Ferrari's method)
% running our coefficients through Mathmetica (quartic_solution.nb)
% optimized to use as few operations as possible...
        
%%% precompute certain terms
v2 = v .* v;
v3 = v2 .* v;
v4 = v3 .* v;
m2 = m .* m;
m3 = m2 .* m;
  
%% Compute alpha & beta
alpha = -1.125*v2;
beta2 = 0.25*v3;
  
%%% Compute p,q,r and u directly.
q = -0.125*(m.*v2);
r1 = -q/2 + sqrt(-m3/27 + (m2.*v4)/256);

u = exp(log(r1)/3); 
y = 2*(-5/18*alpha + u + (m./(3*u))); 
    
W = sqrt(alpha./3 + y);
  
%%% now form all 4 roots
root = zeros(size(v,1),size(v,2),4);
root(:,:,1) = 0.75.*v  +  0.5.*(W + sqrt(-(alpha + y + beta2./W )));
root(:,:,2) = 0.75.*v  +  0.5.*(W - sqrt(-(alpha + y + beta2./W )));
root(:,:,3) = 0.75.*v  +  0.5.*(-W + sqrt(-(alpha + y - beta2./W )));
root(:,:,4) = 0.75.*v  +  0.5.*(-W - sqrt(-(alpha + y - beta2./W )));
  
    
%%%%%% Now pick the correct root, including zero option.
  
%%% Clever fast approach that avoids lookups
v2 = repmat(v,[1 1 4]); 
sv2 = sign(v2);
rsv2 = real(root).*sv2;
    
%%% condensed fast version
%%%             take out imaginary                roots above v/2            but below v
root_flag3 = sort(((abs(imag(root))<epsilon) & ((rsv2)>(abs(v2)/2)) &  ((rsv2)<(abs(v2)))).*rsv2,3,'descend').*sv2;
%%% take best
w=root_flag3(:,:,1);

File: colorfilter/paper/Deconvolution/shrinkv23.py
"""
Explanation:
Precomputation and Setup:
k, m, v2, v3, v4, m2, and m3 are computed to set up the coefficients for the quartic equation.
alpha and beta2 are calculated based on v and beta.
Quartic Equation Roots Calculation:
r1 is computed for solving the cubic equation.
u is computed as the cubic root of r1.
W is computed from alpha and y.
Roots Computation:
Four potential roots are calculated using alpha, beta2, and W.
Selection of the Best Root:
root_flag3 filters out invalid roots and chooses the best one based on criteria.
Return Value:
w is the final result which is selected based on the filtered roots.
Note:

Ensure you have numpy installed in your Python environment. If not, install it using pip install numpy.
The example values for v and beta are placeholders. Replace them with your actual data for practical use.
Adjust the handling of v, especially if it's a multi-dimensional array, based on your specific needs.
"""

import numpy as np

def shrinkv23(v, beta):
    epsilon = 1e-6  # tolerance for imaginary part of real root

    k = 8 / (27 * beta**3)
    m = np.ones_like(v) * k

    # Precompute certain terms
    v2 = v ** 2
    v3 = v2 * v
    v4 = v3 * v
    m2 = m ** 2
    m3 = m2 * m

    # Compute alpha and beta
    alpha = -1.125 * v2
    beta2 = 0.25 * v3

    # Compute p, q, r and u directly
    q = -0.125 * m * v2
    r1 = -q / 2 + np.sqrt(-m3 / 27 + (m2 * v4) / 256)
    
    # Use the cubic root for u
    u = np.exp(np.log(r1) / 3)
    y = 2 * (-5 / 18 * alpha + u + (m / (3 * u)))
    
    W = np.sqrt(alpha / 3 + y)

    # Form all 4 roots
    root = np.zeros((v.shape[0], v.shape[1], 4))
    root[:, :, 0] = 0.75 * v + 0.5 * (W + np.sqrt(-(alpha + y + beta2 / W)))
    root[:, :, 1] = 0.75 * v + 0.5 * (W - np.sqrt(-(alpha + y + beta2 / W)))
    root[:, :, 2] = 0.75 * v + 0.5 * (-W + np.sqrt(-(alpha + y - beta2 / W)))
    root[:, :, 3] = 0.75 * v + 0.5 * (-W - np.sqrt(-(alpha + y - beta2 / W)))

    # Determine the best root
    v2_expanded = np.repeat(v[:, :, np.newaxis], 4, axis=2)
    sv2 = np.sign(v2_expanded)
    rsv2 = np.real(root) * sv2
    
    # Filter roots: take out imaginary roots, those above v/2 but below v
    valid_root = ((np.abs(np.imag(root)) < epsilon) &
                  (rsv2 > np.abs(v2_expanded) / 2) &
                  (rsv2 < np.abs(v2_expanded)))
    root_flag3 = np.sort(valid_root * rsv2, axis=2)[:, :, ::-1] * sv2
    
    # Take the best root
    w = root_flag3[:, :, 0]
    
    return w

# Example usage
v = np.random.rand(100, 100)  # Example input
beta = 0.1  # Example beta value
w = shrinkv23(v, beta)

